code/game/machinery/tesla_turret.dm

```diff
@@ -1,137 +1,619 @@
+
+
+/obj/machinery/tesla_turret/ex_act(severity)
+	switch (severity)
+		if (1)
+			take_damage(rand(140,300))
+		if (2)
+			take_damage(rand(80,170))
+		if (3)
+			take_damage(rand(50,120))
+
+/obj/machinery/tesla_turret/proc/die()	//called when the turret dies, ie, health <= 0
+	health = 0
+	stat |= BROKEN	//enables the BROKEN bit
+	spark_system.start()	//creates some sparks because they look cool
 	update_icon()
 
-/obj/machinery/power/tesla_turret/Process()
-	..()
-	if(active)
-		for(var/mob/M in view(5, src))
-			if(istype(M, current_target))
-				if(M.stat != DEAD)
-					zap(M)
-				else // They're dead, don't shoot
-					continue
-
-/obj/machinery/power/tesla_turret/update_icon()
-	..()
-	if(active)
-		icon_state = "coil_active"
-		set_light(3)
-	else
-		icon_state = "coil"
-		set_light(0)
+/obj/machinery/tesla_turret/Process()
+	//the main machinery process
 
+	if(stat & (NOPOWER|BROKEN))
+		last_target = null
+		return
 
-/obj/machinery/power/tesla_turret/proc/zap(mob/living/target)
-	if((!powernet || last_zap + zap_cooldown) > world.time)
-		return FALSE
-	last_zap = world.time
-	var/power = (powernet.avail/2)
-	draw_power(power)
+	if(!enabled)
+		last_target = null
+		return
+
+	var/list/targets = list()			//list of primary targets
+	var/list/secondarytargets = list()	//targets that are least important
+
+	for(var/mob/M in mobs_in_view(world.view, src))
+		assess_and_assign(M, targets, secondarytargets)
+
+	if(!tryToShootAt(targets))
+		tryToShootAt(secondarytargets) // if no valid targets, go for secondary targets
+
+/obj/machinery/tesla_turret/proc/assess_and_assign(var/mob/living/L, var/list/targets, var/list/secondarytargets)
+	switch(assess_living(L))
+		if(TURRET_PRIORITY_TARGET)
+			targets += L
+		if(TURRET_SECONDARY_TARGET)
+			secondarytargets += L
+
+/obj/machinery/tesla_turret/proc/assess_living(var/mob/living/L)
+	var/obj/item/card/id/id_card = L.GetIdCard()
+
+	if(id_card && (id_card.registered_name in shock_net.registered_names))
+		return TURRET_NOT_TARGET
+
+	if(!istype(L))
+		return TURRET_NOT_TARGET
+
+	if(L.invisibility >= INVISIBILITY_LEVEL_ONE) // Cannot see him. see_invisible is a mob-var
+		return TURRET_NOT_TARGET
+
+	if(L.stat)		//if the perp is dead/dying, no need to bother really
+		return TURRET_NOT_TARGET	//move onto next potential victim!
+
+	if(!L)
+		return TURRET_NOT_TARGET
+	/*
+	if(colony_allied_turret && L.colony_friend) //Dont target colony pets if were allied with them
+		return TURRET_NOT_TARGET
+
+	if(!colony_allied_turret && !L.colony_friend) //If were not allied to the colony we dont attack anything thats against the colony
+		return TURRET_NOT_TARGET
+
+	if(!colony_allied_turret && L.colony_friend) //If were not allied with the colony we attack them and their pets
+		return TURRET_SECONDARY_TARGET
+
+	if(colony_allied_turret && !L.colony_friend) //If were allied with the colony and we attack things that are not are pets
+		return TURRET_SECONDARY_TARGET
+	*/
+
+	if(issilicon(L))
```

Comment:
So it will shoot at synths even if they aren't too far away?

Reply:
Fair enough: I can fix that.